Add a command to list external packages (#24082)

jjolly opened the pull request:
fixes #24082 

This commit adds an option to the:
```
spack external find
```

command that allows it to list already-configured external packages. This allows the display of packages that have external specs without having to manually parse the `packages.yaml` file.

Example of use:

```
% spack external find --in-config
==> The following external specs are defined in /home/jjolly/.spack/packages.yaml
bison@3.0.4  cmake@3.19.1  flex@2.5.35
```

Commit:
external find: Add option to list external packages (#24082)

This commit adds an option to the:

```
spack external find
```

command that allows it to list already-configured external packages. This allows the display of packages that have external specs without having to manually parse the `packages.yaml` file.

Example of use:

```
% spack external find --in-config
==> The following external specs are defined in /home/jjolly/.spack/packages.yaml
bison@3.0.4  cmake@3.19.1  flex@2.5.35
```

haampie commented:
Seems useful, but to me it makes more sense to have this as the default behavior of `spack external list`, and move what is now spack external list to say `spack external find --dry-run` or something. (Not saying you should implement it like this, but opening the discussion...)

jjolly commented:
> Seems useful, but to me it makes more sense to have this as the default behavior of `spack external list`, and move what is now spack external list to say `spack external find --dry-run` or something. (Not saying you should implement it like this, but opening the discussion...)

I like your idea, but my philosophy is "change with the least impact," and switching functionality around makes me gibber a bit. Perhaps for 0.17, right?

alalazo commented:
What if we add:
```console
$ spack external show
```
? In that way we won't modify an existing command and we won't add an option that completely changes the behavior of `spack external find`. Drawback may be that `list/show` can be considered as synonyms, so easy to confuse if somebody doesn't read the help.

Commit:
external show: Add show to external command

This commit adds the show variant to the external command. This variant will show the already-configured external packages found in the packages.yaml file.

Example of use:

```
% spack external show
==> The following external specs are defined in /home/jjolly/.spack/packages.yaml
bison@3.0.4  cmake@3.19.1  flex@2.5.35
```

jjolly commented:
> What if we add:
> 
> ```
> $ spack external show
> ```
> 
> ? In that way we won't modify an existing command and we won't add an option that completely changes the behavior of `spack external find`. Drawback may be that `list/show` can be considered as synonyms, so easy to confuse if somebody doesn't read the help.

Having a `show` variant to `external` is probably a better idiom than add an output-only option to a modifying variant like `list`. I have updated the commit to have the `show` variant. I did not remove the `--in-config` switch patch but just added a commit over it. If you think the commits should be squashed I would be happy to do that.

jjolly commented:
After putting a little more thought into it, perhaps the command should be `spack external config` to display the current configuration of external packages. Thoughts?

alalazo commented:
> After putting a little more thought into it, perhaps the command should be spack external config to display the current configuration of external packages. Thoughts?

We already have something very similar:
```console
$ spack config get packages
```
which I think doesn't need a duplicate. I like that the subcommand in this PR can print a high-level summary of the externals that are configured. In my opinion if a user needs more information than what returned here then `spack config` is the way to go.

alalazo commented:
Moving out of the point release project since deciding whether we might need a command like this one in the long term requires some thought. 

EDIT:

To be less cryptic, having a concretizer that can reuse specs opens the possibility to add detected externals directly into the DB with e.g. missing variants (the ones that were not detectable). We never discussed this possibility in any meeting yet but, if we go in that direction, the best way to achieve the goal that this PR has is to add an option to `spack find` to filter the DB for external specs.

alecbcs commented:
@alalazo @jjolly @haampie is this PR / functionality still something we'd like to pursue or are the commands available now (2+ years later) sufficient?

alecbcs commented:
Closing with no response. Feel free to reopen if folks want to continue working on this.
